### PR TITLE
Run on change

### PR DESCRIPTION
**Pull request by assignUser**
Another unsolicited pr, I hope that's ok?

As I mentioned we are going to rework the core of our package which has a lot of components to change and keep the performance in mind. To make relevant information easier to spot within the commit messages I implemented a function that checks for changes between head and base ref and only executes the benchmark if any of the files passed to the function were modified.

```r
# You can run tests only if the relevant file changed
touchstone::run_on_change("R/utils.R", touchstone::benchmark_run_ref(
  random_test = yourpkg::f(), #<- TODO put the call you want to benchmark here
  n = 2
))
```

`gert::git_diff` does currently not support this, so I had to use `system` to call `git diff ref1 ref2 --` directly (which should not be an issue either locally nor on gha but it would be nicer to use gert... if changes can not be checked the benchmark will be run.

Let me know what you think. :)

### CODE REVIEW

**Comment by lorenzwalthert** [comment]
> Another unsolicited pr, I hope that's ok?

I prefer to first discuss things in an issue. Of course you can spend time on the implementation and commit to your fork, but I can’t guaranteed I’ll accept all PRs 😜 

Also, i am a but time constraint so I can’t guarantee to respond quickly.

**Comment by assignUser** [comment]
No worries, take your time :) I will use my fork for now 😁 

I have some more ideas but I will open issues for them first^^

File: inst/script.R [hunk 1]
```diff
@@ -19,6 +19,11 @@ touchstone::benchmark_run_ref(
...
 )
 
+# You can run tests only if the relevant file changed
+touchstone::run_on_change("R/utils.R", touchstone::benchmark_run_ref(
+  random_test = yourpkg::f(), #<- TODO put the call you want to benchmark here
+  n = 2
+))
 
 # create artifacts used downstream in the GitHub Action
 touchstone::benchmarks_analyze()
```
Comment (lorenzwalthert) [inline inst/script.R:29]: I note this in a few files. Can you please use pre-commit on your prposed changes, there is an end-of-file-fixer that ensures you end with an empty line.

Comment (assignUser) [reply inst/script.R:29]: Ah this is "your fault" ;P as I use r-lsp for formatting in vsc and they use ` styler::tidyverse_style(indent_by = options$tabSize)` 

But I will look into using precommit :)

Comment (lorenzwalthert) [reply inst/script.R:29]: aww yes. Please upvote https://github.com/REditorSupport/languageserver/issues/462.

**Comment by assignUser** [comment]
> > To make relevant information easier to spot within the commit messages
> 
> I am not sure I understand your workflow. Can you elaborate?

I was thinking about the number of benchmarks I am going to use (20+ easy) and how the commit message would look. I am actually not sure if this is still necessary with #43 / #44 , well with large number of iterations and such it could still save some time on the gh action.

**Comment by lorenzwalthert** [comment]
I see where you are coming from, but I don't think this should be implemented as is for the following reasons:

* Instead of a new function (that convolutes the namespace), I think this should be a an argument to `benchmark_run_ref()`. like this:
```r
touchstone::benchmark_run_ref(
  random_test = yourpkg::f(), #<- TODO put the call you want to benchmark here
  n = 2, skip = has_source_change('R/sample.R')
)
```
* Then, I think referencing a file for skipping is quite brittle because of reverse dependencies. E.g. your code depends on a function in `R/sample.R`, which in turn depends on another function in another file. Your suggestion (as I understand it) will have a false negative here. I am only willing to support skipping if we can get it right in general. There are a few tools out there for static code analysis, but a quick search did not help me to give me the solution I wanted: A package from CRAN that, given a function in a package, gives me all reverse dependencies. Then, we'd have to hash their body and formals to create a cache so we can check if any of this changed. A few hints: {codetools}, {CodeDepends}, {mvbutils}, but they all don't have great documentation. Or https://stackoverflow.com/questions/8761857/identifying-dependencies-of-r-functions-and-scripts.
* Maybe there is an other way of inferring the reverse dependencies, e.g. with extending my experimental https://github.com/lorenzwalthert/parsesum (which you starred at some point), but this is still very experimental, not on CRAN and has probably 0 tests.
* If you can figure it out, I am happy to proceed with this, otherwise, I don't think I'd want this addition.
* This functionality seems most relevant for local benchmarking. The focus on {touchstone} was on CI/CD runs (which can be extended, I am not in general against it).

**Comment by assignUser** [comment]
I like the `skip` argument, much cleaner. I also agree with the rest of your points, my viewpoint was a bit narrow and focused on my use case, thank you for your input!

My implementation is indeed very brittle, even against refactoring (which will probably happen in my case, so I did not even think that through properly). Thanks for your thoughts and preliminary research on this. :)

I was following the spirit of your `Cancel` action in avoiding unnecessary run time. I think this could still work, it is just a lot more complicated than my initial idea. I converted to draft for now.

**Comment by lorenzwalthert** [comment]
Oh and this RStudio employee has written a package to import functions from a package without declaring it as import (basically copying source). I think he must also do a dependency analysis to avoid copying the whole package. https://github.com/wch/staticimports

**Comment by lorenzwalthert** [comment]
Maybe https://mobile.twitter.com/antoine_fabri/status/1439348903770472452

**Pull request closed without merging**